### ISSUE TITLE
Trusted registries don't work correctly for multi-part repo names

### ISSUE THREAD

**Issue by benclive** [hadolint/hadolint#401]
- [x] This is a bug report
- [ ] This is a feature request
- [ ] I searched existing issues before opening this one

Images such as "gcr.io/distroless/base-debian10" are hosted in the "gcr.io/distroless" registry. Hadolint only checks the "gcr.io" part of the registry against the trusted registry list.

Private Google Container Registry (GCR) registries are of the form "gcr.io/\<google project name\>" so trusting the whole of gcr.io is a security risk.

### Expected behavior
Trusting "gcr.io/distroless" for image "gcr.io/distroless/base-debian10" should not return an error.
Trusting "gcr.io" for image "gcr.io/distroless/base-debian10" should return an untrusted error.

### Actual behavior
Trusting "gcr.io/distroless" for image "gcr.io/distroless/base-debian10" returns an error.
Trusting "gcr.io" for image "gcr.io/distroless/base-debian10" does not return an error.

### Steps to reproduce the behavior
Using a minimal Dockerfile:
```
FROM gcr.io/distroless/base-debian10:latest
```

#### Case 1

Running
```
hadolint --trusted-registry gcr.io/distroless Dockerfile
```
returns
```
Dockerfile:1 DL3007 Using latest is prone to errors if the image will ever update. Pin the version explicitly to a release tag
Dockerfile:1 DL3026 Use only an allowed registry in the FROM image
```
where DL3026 is unexpected.

#### Case 2
Running 
```
hadolint --trusted-registry gcr.io Dockerfile
```
returns
```
Dockerfile:1 DL3007 Using latest is prone to errors if the image will ever update. Pin the version explicitly to a release tag
```
only which is unexpected as **DL3026** is missing.

**Output of `hadolint --version` or
  `docker run --rm hadolint/hadolint hadolint --version`:**

```bash
$ hadolint --version
Haskell Dockerfile Linter 1.17.4-no-git
```

**Dockerfile** (if relevant)

```Dockerfile
FROM gcr.io/distroless/base-debian10:latest
```

**Additional environment details (OS, stack version, etc.)**
I'm running on a Mac and installed via Homebrew. This happens from both the command line and using a yaml configuration file.

**Comment by lorenzo** [comment]
We currently interpret the registry as the `domain` + `port` part of an image. After that, the rest of the name is considered the image name. This is so it can work for things like `hub.docker.com/nginx` but also `hub.docker.com/user/repo`, which is what most people would expect (including myself).

An enhancement that could be done is to restring the allowed images to a certain pattern.

**Comment by ghostsquad** [comment]
I don't see a reason why you can't do a full match on the registry as compared to what's in the FROM statement. This should be backwards compatible. If someone only trusts `mycompany/nginx`, then that's what they use in the option. Otherwise, they would only use `mycompany`

**Comment by setpill** [comment]
I agree that

> Trusting "gcr.io/distroless" for image "gcr.io/distroless/base-debian10" should not return an error.

I disagree that

> Trusting "gcr.io" for image "gcr.io/distroless/base-debian10" should return an untrusted error.

That latter one requires hardcoding of `gcr.io` behaviour which doesn't scale. If the user blanket trusts `gcr.io` why SHOULDN'T hadolint accept that?

**Comment by 19D** [comment]
> I disagree that
> 
> > Trusting "gcr.io" for image "gcr.io/distroless/base-debian10" should return an untrusted error.
> 
> That latter one requires hardcoding of `gcr.io` behaviour which doesn't scale. If the user blanket trusts `gcr.io` why SHOULDN'T hadolint accept that?

this is correct. I would like the ability to filter trusted registries by adding further url paths like `trustedregistry.com/approvedImages/*`

**Comment by ringerc** [comment]
Possibly related to https://github.com/hadolint/hadolint/issues/355

**Comment by arayofcode** [comment]
Any plans to fix this?

**Comment by Mohataseem89** [comment]
i would like to work on this after reading the source here is what is happening and what I wouldd propose.

**Root cause**
`parseRegistry` in `language-docker` stops at the first `/`, so `FROM gcr.io/distroless/base-debian10` is parsed as:
- `registryName = "gcr.io"`
- `imageName = "distroless/base-debian10"`

The `/distroless` org path ends up inside `imageName`, never in `registryName`. So `DL3026` can only ever see `"gcr.io"` — matching `"gcr.io/distroless"` is currently impossible.

**Proposed fix (rule-layer only, no parser changes)**
Instead of only passing `registryName` to `isRegistryAllowed`, reconstruct the full image path and do a prefix match:

```haskell
isAllowed Image {registryName = Just registry, imageName} =
  isRegistryAllowed (unRegistry registry <> "/" <> imageName)
    || isRegistryAllowed (unRegistry registry)
```

This means:
- `--trusted-registry gcr.io/distroless`  matches `gcr.io/distroless/base-debian10` (prefix match on full path)
- `--trusted-registry gcr.io`  still matches `gcr.io/distroless/base-debian10` (fallback to registry-only match)

**One question before I open the PR:** the original reporter also wanted `--trusted-registry gcr.io` to *fail* for `gcr.io/distroless/base-debian10`, but several commenters disagreed. My proposal keeps `gcr.io` as a valid blanket trust. Is that the intended behaviour, or should the match be strict (exact prefix only, no fallback)?

Happy to go either way once confirmed.
